docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,233 @@
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF"
+            }
+        ],
+        "type": "library"
+    },
+    {
+        "name": "Llama-3.1-Nemotron-70B-Instruct-HF",
+        "purl": "pkg:huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF@043235d6088ecd3dd5fb5ca3592b6913fd516027?checksum=sha256:e4ca5fad&file_name=model-00001-of-00030.safetensors",
+        "properties": [
+            {
+                "name": "cachi2:found_by",
+                "value": "cachi2:generic"
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/fac73d3507320ec1258620423469b4b38f88df6e/model-00001-of-00030.safetensors"
+            }
+        ],
+        "type": "file"
+    }
+]
+```
+
+### OCI
+
+#### generic_artifacts.yaml
+```yaml
+metadata:
+    version: '1.0'
+artifacts:
+  - purl: pkg:oci/buildah-task@sha256%3Ab2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c?repository_url=quay.io/konflux-ci&arch=amd64&tag=latest
+    target: quakus.jar
+```
+
+#### resolving the download url
+```python
+# parse the input purl
+pkg:oci/{name}@{algorithm}%3A{digest}?repository_url={repository_url_and_namespace}[&arch={arch}][&tag={tag}]
+
+# generate the download command
+podman pull {repository_url_and_namespace}/{name}@{algorithm}:{digest}
+```
+
+#### SBOM component
+```json
+{
+    "name": "buildah-task",
+    "purl": "pkg:oci/buildah-task@sha256%3Ab2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c?repository_url=quay.io/konflux-ci&arch=amd64&tag=latest",
+    "properties": [
+        {
+            "name": "cachi2:found_by",
+            "value": "cachi2:generic"
+        }
+    ],
+    "externalReferences": [
+        {
+            "type": "distribution",
+            "url": "quay.io/konflux-ci/buildah-task@sha256:b2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c"
+        }
+    ],
+    "type": "container"
+}
+```
+
+## Initial thoughts
+
+From a Cachi2 perspective, we can separate purls types into ones that are part of existing package manager ecosystems (such as nuget, composer, maven) and ones that are not (github, huggingface, oci):
+- Types that refer to existing package managers seem to have a straightforward resolution. The package is likely a file in a predictable URL within a registry.
+- Types that don't refer to existing package managers are trickier to resolve, since they might be poiting to a container image or a full git repository. They might also need extra tools to provide a smooth resolution (podman, git, etc)
+
+
+### Decision points
+
+- Should the checksums be specified as part of the input purl?
+- Should we limit the qualifiers to the types that are strictly available to that type?
```

Comment:
Yes, the purl specification loosely determines what are the qualifiers for each type. See [oci,](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#oci) for example. 

There are also global qualifiers that can be applied to any type (check the qualifiers bullet [here](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#rules-for-each-purl-component)).

Reply:
I think I would expect the qualifiers to be verified as part of the parser. Cachi2 would need to ensure that the required qualifiers are present and that the optional ones may be present.

Which of the package-specific qualifiers that are directly specified would Cachi2 use? For example, would OCI artifact support be able to handle the repository_url, namespace, version, ... or would the input purl provide those _plus_ a download_url? Or would support for these qualifiers be considered more for "full" package manager support instead of generic?

Reply:
I believe the specification for [oci](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#oci) seems to indicate the use of `repository_url`.

I changed the design so that it mentions we will validate the qualifiers depending on type, it is probably better than to leave it completely open.

Reply:
I haven't read the content that you have changed yet, but that change makes sense.